dpath/segments.py

```diff
@@ -0,0 +1,386 @@
+    else:
+        leaves = (bytes, str, int, float, bool, type(None))
+    return isinstance(thing, leaves)
+
+
+def leafy(thing):
+    '''
+    Same as leaf(thing), but also treats empty sequences and
+    dictionaries as True.
+
+    leafy(thing) -> bool
+    '''
+    return leaf(thing) or len(thing) == 0
+
+
+def walk(obj, location=()):
+    '''
+    Yield all valid (segments, value) pairs (from a breadth-first
+    search, right-to-left on sequences).
+
+    walk(obj) -> (generator -> (segments, value))
+    '''
+    if not leaf(obj):
+        for k, v in kvs(obj):
+            length = None
+
+            try:
+                length = len(k)
+            except:
+                pass
+
+            if length is not None and length == 0 and not options.ALLOW_EMPTY_STRING_KEYS:
+                raise InvalidKeyName("Empty string keys not allowed without "
+                                     "dpath.options.ALLOW_EMPTY_STRING_KEYS=True: "
+                                     "{}".format(location + (k,)))
+            yield ((location + (k,)), v)
+        for k, v in kvs(obj):
+            for found in walk(v, location + (k,)):
+                yield found
+
+
+def get(obj, segments):
+    '''
+    Return the value at the path indicated by segments.
+
+    get(obj, segments) -> value
+    '''
+    current = obj
+    for (i, segment) in enumerate(segments):
+        if leaf(current):
+            raise PathNotFound('Path: {}[{}]'.format(segments, i))
+
+        current = current[segment]
+    return current
+
+
+def has(obj, segments):
+    '''
+    Return True if the path exists in the obj. Otherwise return False.
+
+    has(obj, segments) -> bool
+    '''
+    try:
+        get(obj, segments)
+        return True
+    except:
+        return False
+
+
+def expand(segments):
+    '''
+    Yield a tuple of segments for each possible length of segments.
+    Starting from the shortest length of segments and increasing by 1.
+
+    expand(keys) -> (..., keys[:-2], keys[:-1])
+    '''
+    index = 0
+    for segment in segments:
+        index += 1
+        yield segments[:index]
+
+
+def types(obj, segments):
+    '''
+    For each segment produce a tuple of (segment, type(value)).
+
+    types(obj, segments) -> ((segment[0], type0), (segment[1], type1), ...)
+    '''
+    result = []
+    for depth in expand(segments):
+        result.append((depth[-1], type(get(obj, depth))))
+    return tuple(result)
+
+
+def leaves(obj):
+    '''
+    Yield all leaves as (segment, value) pairs.
+
+    leaves(obj) -> (generator -> (segment, value))
+    '''
+    return ifilter(lambda p: leafy(p[1]), walk(obj))
+
+
+def int_str(segment):
+    '''
+    If the segment is an integer, return the string conversion.
+    Otherwise return the segment unchanged. The conversion uses 'str'
+    which means that:
+
+        * in Python 2.x: 0 -> b'0'
+        * in Python 3.x: 0 -> u'0'
+
+    int_str(segment) -> str
+    '''
+    if isinstance(segment, (int, long)):
```

Comment:
While it's incredibly unlikely that someone will subclass int/long, would it be worth it to change these to issubclass() calls (both here and throughout)? Our bugtracker is full of edge cases lately, and this one seems easy to head off at the pass.